src/phpDocumentor/Dsn.php

```diff
@@ -0,0 +1,232 @@
+<?php
+/**
+ * This file is part of phpDocumentor.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @copyright 2010-2015 Mike van Riel<mike@phpdoc.org>
+ * @license   http://www.opensource.org/licenses/mit-license.php MIT
+ * @link      http://phpdoc.org
+ */
+
+namespace phpDocumentor;
+
+/**
+ * Value Object for DSN.
+ */
+final class Dsn
+{
+    /** @var string */
+    private $scheme;
+
+    /** @var string */
+    private $host;
+
+    /** @var int */
+    private $port;
+
+    /** @var string */
+    private $user;
+
+    /** @var string */
+    private $password;
+
+    /** @var string */
+    private $path;
+
+    /** @var string[]  */
+    private $query = [];
+
+    /** @var string[]  */
+    private $parameters = [];
+
+    /**
+     * Initializes the Dsn
+     *
+     * @param string $dsn
+     */
+    public function __construct($dsn)
+    {
+        $this->parse($dsn);
+    }
+
+    /**
+     * Returns the scheme part of the DSN
+     *
+     * @return string
+     */
+    public function getScheme()
+    {
+        return $this->scheme;
+    }
+
+    /**
+     * Returns the host part of the DSN
+     *
+     * @return string
+     */
+    public function getHost()
+    {
+        return $this->host;
+    }
+
+    /**
+     * Returns the port part of the DSN
+     *
+     * @return int
+     */
+    public function getPort()
+    {
+        return $this->port;
+    }
+
+    /**
+     * Returns the username part of the DSN
+     *
+     * @return string
+     */
+    public function getUsername()
+    {
+        return $this->user;
+    }
+
+    /**
+     * Returns the password part of the DSN
+     *
+     * @return string
+     */
+    public function getPassword()
+    {
+        return $this->password;
+    }
+
+    /**
+     * Returns the path part of the DSN
+     *
+     * @return string
+     */
+    public function getPath()
+    {
+        return new Path($this->path);
+    }
+
+    /**
+     * Returns the query part of the DSN
+     *
+     * @return string[]
+     */
+    public function getQuery()
+    {
+        return $this->query;
+    }
+
+    /**
+     * Returns the parameters part of the DSN
+     *
+     * @return string[]
+     */
+    public function getParameters()
+    {
+        return $this->parameters;
+    }
+
+    /**
+     * Parses the given DSN
+     *
+     * @param string $dsn
+     */
+    private function parse($dsn)
+    {
+        if (! is_string($dsn)) {
+            throw new \InvalidArgumentException(
+                sprintf('"%s" is not a valid DSN.', $dsn)
```

Comment:
This message won't work; since $dsn is not a string you cannot include it in a sprintf using %s; one solution would be to wrap it in a `var_export($dsn, true)` call. Using var_export with the second flag enabled will return a string, independent of the type of object